aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -1552,47 +1646,131 @@ impl AptosVM {
-                // the loader cache. It only puts the module data in the data cache.
-                session.publish_module_bundle_with_compat_config(
-                    bundle.into_inner(),
-                    destination,
-                    gas_meter,
-                    &DummyCodeStorage,
-                    Compatibility::new(
-                        true,
-                        !self
-                            .features()
-                            .is_enabled(FeatureFlag::TREAT_FRIEND_AS_PRIVATE),
-                    ),
-                )?;
+                    Ok(Some(tmp_module_storage.destroy()))
+                } else {
+                    // Validate the module bundle
+                    self.validate_publish_request(
+                        session,
+                        modules,
+                        expected_modules,
+                        allowed_deps,
+                    )?;
 
-                self.execute_module_initialization(
-                    session,
-                    gas_meter,
-                    modules,
-                    exists,
-                    &[destination],
-                    new_published_modules_loaded,
-                    traversal_context,
-                )?;
+                    // Check what modules exist before publishing.
+                    let mut exists = BTreeSet::new();
+                    for m in modules {
+                        let id = m.self_id();
+                        #[allow(deprecated)]
+                        if session.exists_module(&id)? {
+                            exists.insert(id);
+                        }
+                    }
+
+                    // Publish the bundle and execute initializers
+                    // publish_module_bundle doesn't actually load the published module into
+                    // the loader cache. It only puts the module data in the data cache.
+                    #[allow(deprecated)]
+                    session.publish_module_bundle_with_compat_config(
+                        bundle.into_inner(),
+                        destination,
+                        gas_meter,
+                        module_storage,
+                        compat,
+                    )?;
+
+                    self.execute_module_initialization(
+                        session,
+                        module_storage,
+                        gas_meter,
+                        modules,
+                        exists,
+                        &[destination],
+                        new_published_modules_loaded,
+                        traversal_context,
+                    )?;
+                    Ok(None)
+                }
+            } else {
+                Ok::<_, VMError>(
+                    if self.features().use_loader_v2() {
+                        Some(ModuleWriteSet::empty())
+                    } else {
+                        None
+                    },
+                )
```

Comment:
This method is very long, can we extract the above v1/v2 fragment into a helper?

Reply:
Let's couple this with a general cleanup :)